RavenDB-22498 auto to static index converter

ppekrol opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22498

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [x] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

ppekrol committed:
RavenDB-22498 started working on auto to static index converter

ppekrol committed:
RavenDB-22498 implemented /databases/*/indexes/auto/convert endpoint

ppekrol committed:
RavenDB-22498 added test

ppekrol committed:
RavenDB-22498 properly handle complex auto index fields with multiple indexing capabilities

ppekrol committed:
RavenDB-22498 added Result for map reduce

ppekrol committed:
RavenDB-22498 minor

ppekrol committed:
RavenDB-22498 added more auto indexes for testing

ppekrol committed:
RavenDB-22498 skip empty auto indexes, fix group by field names

ppekrol committed:
RavenDB-22498
- support for @all_docs and @empty
- properly detecting AutoFieldIndexing.No

ppekrol committed:
RavenDB-22498 handling constants and nulls in group by

arekpalinski commented:
CC @DejanMilicic

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+        }
+    }
+
+    private enum ConvertType
```
`ConversionOutputType` maybe

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+
+                    csharpClassName = $"{csharpClassName}.cs";
+
+                    HttpContext.Response.Headers[Constants.Headers.ContentDisposition] = $"attachment; filename=\"{csharpClassName}\"; filename*=UTF-8''{csharpClassName}";
```
Maybe introduce `?download=true`?

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+
+        if (Enum.TryParse(typeAsString, ignoreCase: true, out ConvertType convertType) == false)
+            throw new InvalidOperationException($"Could not parse '{typeAsString}' to any known conversion type.");
+
```
I wonder whether we should introduce `?language=` parameter. Currently we'll support only C# syntax and throw on JavaScript

ppekrol replied:
Let's leave it, but I would call it 'syntax' or sth like that

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+    private enum ConvertType
+    {
+        Csharp,
+        Export
```
Maybe `Json`, `JsonDefinition`, `PotrableJson` ?

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+
+    private enum ConvertType
+    {
+        Csharp,
```
`Csharp_class` maybe to avoid the confusion with `CSharp` as the language parameter?

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+                        options.Storage = FieldStorage.Yes;
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
```
Shouldn't we handle `FieldStorage.No` ? Even if it's a no op

arekpalinski replied:
And please provide message to the exception

ppekrol replied:
No, this generates a lot of noise, for each field in the definition then it is set to 'No', same for AbstractIndexCreationTask, you are getting `fields x Store(name, No)`

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+        static void ConstructMap(AutoIndexDefinition autoIndex, StringBuilder sb, AutoIndexConversionContext context)
+        {
+            sb
+                .AppendLine("Map = items => from item in items")
```
Don't we need to specify collection name e.g. `in docs.Items`? where `Items` is a given collection name

ppekrol replied:
no, this is LINQ syntax

arekpalinski commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+                                sb.Append($"{nameof(GeographySpatialOptionsFactory.BoundingBoxIndex)}({typeof(SpatialUnits).FullName}.{spatial.Units})");
+                                break;
+                            default:
+                                throw new ArgumentOutOfRangeException();
```
Please provide the argument name and the message to all `ArgumentOutOfRangeException`

arekpalinski reviewed (commented):
1. In auto indexes we have the notion of `GroupByArrayBehavior`. I didn't see any handling of that, do we need to handle it explicitly?

2. What about the handling of `AutoIndexField.HasQuotedName`?

Lwiel commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+                    options.Storage = FieldStorage.Yes;
+                    options.TermVector = FieldTermVector.WithPositionsAndOffsets;
+                    return;
+                }
```
What about `AutoFieldIndexing.Default`? And what if we have multiple flags set?

ppekrol replied:
since this is a default then there is nothing to set

Lwiel commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+                        case AggregationOperation.Sum:
+                        case AggregationOperation.Count:
+                            {
+                                sb.AppendLine($"public int {f.FieldName} {{ get; set; }}");
```
What about sum of floating point numbers? And would a single pair of brackets be enough?

ppekrol replied:
I do not know anything about the type, so wild guess here. User will have to make the correction if needed.

Lwiel commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+                else
+                {
+                    if (seenConst)
+                        throw new InvalidOperationException("Cannot generate complex group by because const was encountered.");
```
What do you think about rephrasing it to "because a constant field was encountered."?

Lwiel commented on test/SlowTests/Issues/RavenDB_22498.cs:
```diff
@@ -0,0 +1,270 @@
+
+namespace SlowTests.Issues;
+
+public class RavenDB_22498 : RavenTestBase
```
Maybe we could test group by using complex field too?

ppekrol replied:
I'm pretty sure it is part of the slow or fast tests

maciejaszyk commented on src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs:
```diff
@@ -0,0 +1,716 @@
+            : Inflector.Singularize(autoIndex.Collection);
+
+        sb
+            .Append($"public class {csharpClassName} : {typeof(AbstractIndexCreationTask).FullName}<{className}");
+
```
```suggestion
            .Append($"public class {csharpClassName} : {typeof(AbstractIndexCreationTask).FullName}<{className}>");

```

maciejaszyk commented:
Overall LGTM

ppekrol committed:
RavenDB-22498 added exception messages

ppekrol committed:
RavenDB-22498
- renamed ConvertType to ConversionOutputType
- renamed 'type' query string parameter to outputType
- introduced 'download' query string parameter
- fixed ex message

ppekrol committed:
RavenDB-22498 better assertions in tests

ppekrol commented:
Done
